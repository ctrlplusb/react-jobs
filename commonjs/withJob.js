'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.default = withJob;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var validSSRModes = ['resolve', 'defer', 'boundary'];
var neverWorkAgain = function neverWorkAgain() {
  return false;
};

function withJob(config) {
  if ((typeof config === 'undefined' ? 'undefined' : _typeof(config)) !== 'object') {
    throw new Error('You must provide a config object to withJob');
  }

  var work = config.work,
      LoadingComponent = config.LoadingComponent,
      ErrorComponent = config.ErrorComponent,
      _config$serverMode = config.serverMode,
      serverMode = _config$serverMode === undefined ? 'resolve' : _config$serverMode,
      _config$shouldWorkAga = config.shouldWorkAgain,
      shouldWorkAgain = _config$shouldWorkAga === undefined ? neverWorkAgain : _config$shouldWorkAga;


  if (typeof work !== 'function') {
    throw new Error('You must provide a work function to withJob');
  }

  if (validSSRModes.indexOf(serverMode) === -1) {
    throw new Error('Invalid serverMode provided to asyncComponent');
  }

  var env = typeof window === 'undefined' ? 'node' : 'browser';

  return function wrapWithJob(WrappedComponent) {
    var id = void 0;

    var ComponentWithJob = function (_Component) {
      _inherits(ComponentWithJob, _Component);

      function ComponentWithJob(props, context) {
        _classCallCheck(this, ComponentWithJob);

        // Each instance needs it's own id as that is how we expect work to
        // be executed.  It is not shared between element instances.
        var _this = _possibleConstructorReturn(this, (ComponentWithJob.__proto__ || Object.getPrototypeOf(ComponentWithJob)).call(this, props, context));

        _initialiseProps.call(_this);

        if (context.jobs) {
          id = context.jobs.getNextId();
        }
        return _this;
      }

      // @see react-async-bootstrapper


      _createClass(ComponentWithJob, [{
        key: 'asyncBootstrap',
        value: function asyncBootstrap() {
          if (env === 'browser') {
            // No logic for browser, just continue
            return true;
          }

          // node
          return serverMode === 'defer' ? false : this.resolveWork(this.props);
        }
      }, {
        key: 'componentWillMount',
        value: function componentWillMount() {
          var result = void 0;

          if (this.context.jobs) {
            result = env === 'browser' ? this.context.jobs.getRehydrate(id) : this.context.jobs.get(id);
          }

          this.setState({
            data: result ? result.data : null,
            error: result ? result.error : null,
            completed: result != null
          });
        }
      }, {
        key: 'componentDidMount',
        value: function componentDidMount() {
          if (!this.state.completed) {
            this.resolveWork(this.props);
          }

          if (this.context.jobs && env === 'browser') {
            this.context.jobs.removeRehydrate(id);
          }
        }
      }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
          this.unmounted = true;
        }
      }, {
        key: 'componentWillReceiveProps',
        value: function componentWillReceiveProps(nextProps) {
          if (shouldWorkAgain((0, _utils.propsWithoutInternal)(this.props), (0, _utils.propsWithoutInternal)(nextProps), this.getJobState())) {
            this.resolveWork(nextProps);
          }
        }
      }, {
        key: 'render',
        value: function render() {
          var _state = this.state,
              data = _state.data,
              error = _state.error,
              completed = _state.completed;

          if (error) {
            return ErrorComponent ? _react2.default.createElement(ErrorComponent, _extends({}, this.props, { error: error })) : null;
          }
          if (!completed) {
            return LoadingComponent ? _react2.default.createElement(LoadingComponent, this.props) : null;
          }
          return _react2.default.createElement(WrappedComponent, _extends({}, this.props, { jobResult: data }));
        }
      }]);

      return ComponentWithJob;
    }(_react.Component);

    ComponentWithJob.displayName = 'WithJob(' + (0, _utils.getDisplayName)(WrappedComponent) + ')';
    ComponentWithJob.contextTypes = {
      jobs: _propTypes2.default.shape({
        getNextId: _propTypes2.default.func.isRequired,
        register: _propTypes2.default.func.isRequired,
        get: _propTypes2.default.func.isRequired,
        getRehydrate: _propTypes2.default.func.isRequired,
        removeRehydrate: _propTypes2.default.func.isRequired
      })
    };

    var _initialiseProps = function _initialiseProps() {
      var _this2 = this;

      this.resolveWork = function (props) {
        var workDefinition = void 0;

        _this2.setState({ completed: false, data: null, error: null });

        try {
          workDefinition = work(props);
        } catch (error) {
          _this2.setState({ completed: true, error: error });
          // Ensures asyncBootstrap stops
          return false;
        }

        if ((0, _utils.isPromise)(workDefinition)) {
          // Asynchronous result.
          return workDefinition.then(function (data) {
            if (_this2.unmounted) {
              return undefined;
            }
            _this2.setState({ completed: true, data: data });
            if (_this2.context.jobs) {
              _this2.context.jobs.register(id, { data: data });
            }
            // Ensures asyncBootstrap continues
            return true;
          }).catch(function (error) {
            if (_this2.unmounted) {
              return undefined;
            }
            if (env === 'browser') {
              setTimeout(function () {
                if (!_this2.unmounted) {
                  _this2.setState({ completed: true, error: error });
                }
              }, 16);
            } else {
              // node
              // We will at least log the error so that user isn't completely
              // unaware of an error occurring.
              // eslint-disable-next-line no-console
              console.warn('Failed to resolve job');
              // eslint-disable-next-line no-console
              console.warn(error.message);
              // eslint-disable-next-line no-console
              console.warn(error.stack);
              if (_this2.context.jobs) {
                _this2.context.jobs.register(id, {
                  error: {
                    message: error.message,
                    stack: error.stack
                  }
                });
              }
            }
            // Ensures asyncBootstrap stops
            return false;
          });
        }

        // Synchronous result.
        _this2.setState({ completed: true, data: workDefinition, error: null });

        // Ensures asyncBootstrap continues
        return true;
      };

      this.getJobState = function () {
        return {
          completed: _this2.state.completed,
          error: _this2.state.error,
          data: _this2.state.data
        };
      };
    };

    return ComponentWithJob;
  };
}