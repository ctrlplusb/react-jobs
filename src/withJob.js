import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { getDisplayName, isPromise, propsWithoutInternal } from './utils'

const validSSRModes = ['resolve', 'defer', 'boundary']
const neverWorkAgain = () => false

export default function withJob(config) {
  if (typeof config !== 'object') {
    throw new Error('You must provide a config object to withJob')
  }

  const {
    work,
    LoadingComponent,
    ErrorComponent,
    serverMode = 'resolve',
    shouldWorkAgain = neverWorkAgain,
  } = config

  if (typeof work !== 'function') {
    throw new Error('You must provide a work function to withJob')
  }

  if (validSSRModes.indexOf(serverMode) === -1) {
    throw new Error('Invalid serverMode provided to asyncComponent')
  }

  const env = typeof window === 'undefined' ? 'node' : 'browser'

  return function wrapWithJob(WrappedComponent) {
    let id

    class ComponentWithJob extends Component {
      static displayName = `WithJob(${getDisplayName(WrappedComponent)})`

      static contextTypes = {
        jobs: PropTypes.shape({
          getNextId: PropTypes.func.isRequired,
          register: PropTypes.func.isRequired,
          get: PropTypes.func.isRequired,
          getRehydrate: PropTypes.func.isRequired,
          removeRehydrate: PropTypes.func.isRequired,
        }),
      }

      constructor(props, context) {
        super(props, context)

        // Each instance needs it's own id as that is how we expect work to
        // be executed.  It is not shared between element instances.
        if (context.jobs) {
          id = context.jobs.getNextId()
        }
      }

      // @see react-async-bootstrapper
      asyncBootstrap() {
        if (env === 'browser') {
          // No logic for browser, just continue
          return true
        }

        // node
        return serverMode === 'defer' ? false : this.resolveWork(this.props)
      }

      componentWillMount() {
        let result

        if (this.context.jobs) {
          result =
            env === 'browser'
              ? this.context.jobs.getRehydrate(id)
              : this.context.jobs.get(id)
        }

        this.setState({
          data: result ? result.data : null,
          error: result ? result.error : null,
          completed: result != null,
        })
      }

      componentDidMount() {
        if (!this.state.completed) {
          this.resolveWork(this.props)
        }

        if (this.context.jobs && env === 'browser') {
          this.context.jobs.removeRehydrate(id)
        }
      }

      componentWillUnmount() {
        this.unmounted = true
      }

      componentWillReceiveProps(nextProps) {
        if (
          shouldWorkAgain(
            propsWithoutInternal(this.props),
            propsWithoutInternal(nextProps),
            this.getJobState(),
          )
        ) {
          this.resolveWork(nextProps)
        }
      }

      resolveWork = props => {
        let workDefinition

        this.setState({ completed: false, data: null, error: null })

        try {
          workDefinition = work(props)
        } catch (error) {
          this.setState({ completed: true, error })
          // Ensures asyncBootstrap stops
          return false
        }

        if (isPromise(workDefinition)) {
          // Asynchronous result.
          return workDefinition
            .then(data => {
              if (this.unmounted) {
                return undefined
              }
              this.setState({ completed: true, data })
              if (this.context.jobs) {
                this.context.jobs.register(id, { data })
              }
              // Ensures asyncBootstrap continues
              return true
            })
            .catch(error => {
              if (this.unmounted) {
                return undefined
              }
              if (env === 'browser') {
                setTimeout(() => {
                  if (!this.unmounted) {
                    this.setState({ completed: true, error })
                  }
                }, 16)
              } else {
                // node
                // We will at least log the error so that user isn't completely
                // unaware of an error occurring.
                // eslint-disable-next-line no-console
                console.warn('Failed to resolve job')
                // eslint-disable-next-line no-console
                console.warn(error.message)
                // eslint-disable-next-line no-console
                console.warn(error.stack)
                if (this.context.jobs) {
                  this.context.jobs.register(id, {
                    error: {
                      message: error.message,
                      stack: error.stack,
                    },
                  })
                }
              }
              // Ensures asyncBootstrap stops
              return false
            })
        }

        // Synchronous result.
        this.setState({ completed: true, data: workDefinition, error: null })

        // Ensures asyncBootstrap continues
        return true
      }

      getJobState = () => ({
        completed: this.state.completed,
        error: this.state.error,
        data: this.state.data,
      })

      render() {
        const { data, error, completed } = this.state
        if (error) {
          return ErrorComponent ? (
            <ErrorComponent {...this.props} error={error} />
          ) : null
        }
        if (!completed) {
          return LoadingComponent ? <LoadingComponent {...this.props} /> : null
        }
        return <WrappedComponent {...this.props} jobResult={data} />
      }
    }

    return ComponentWithJob
  }
}
