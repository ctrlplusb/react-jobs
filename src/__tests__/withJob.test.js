/* eslint-disable react/prop-types */

import React from 'react'
import { mount } from 'enzyme'
import {
  resolveAfter,
  rejectAfter,
  warningsToErrors,
} from '../lib/test-helpers'
import withJob from '../withJob'

const workTime = 10 // ms

function ResultRenderer({ jobResult, children }) {
  return (
    <div>
      {jobResult}
      {children && <div>{children}</div>}
    </div>
  )
}
const ErrorComponent = ({ error }) => <div>{error ? error.message : null}</div>
const LoadingComponent = () => <div>Loading...</div>

fdescribe('withJob()', () => {
  warningsToErrors()

  describe('arguments', () => {
    it('returns a function', () => {
      const actual = typeof withJob({ work: () => undefined })
      const expected = 'function'
      expect(actual).toEqual(expected)
    })

    it('should throws if no config is provided', () => {
      expect(() => withJob()).toThrowError(
        'You must provide a config object to withJob',
      )
    })

    it('should throws if the work is invalid', () => {
      // $FlowIgnore: we expect this to flow error
      expect(() => withJob({ work: 1 })).toThrowError(
        'You must provide a work function to withJob',
      )
    })
  })

  describe('higher order component', () => {
    const hoc = withJob({ work: () => resolveAfter(1) })
    const Actual = hoc(ResultRenderer)

    it('should return a renderable component', () => {
      expect(() => mount(<Actual />)).not.toThrowError()
    })
  })

  describe('rendering', () => {
    it('should set the "result" immediately if the work does not return a promise', () => {
      const ResultRendererWithJob = withJob({ work: () => 'bob' })(
        ResultRenderer,
      )
      expect(mount(<ResultRendererWithJob />)).toMatchSnapshot()
    })

    it('should set the "result" immediately if the work return null', () => {
      const ResultRendererWithJob = withJob({ work: () => null })(
        ResultRenderer,
      )
      expect(mount(<ResultRendererWithJob />)).toMatchSnapshot()
    })

    it('should provide the props to the work function', () => {
      const expected = { foo: 'bar', baz: 'qux' }
      let actual
      const ResultRendererWithJob = withJob({
        work: props => {
          actual = props
        },
      })(ResultRenderer)
      mount(<ResultRendererWithJob {...expected} />)
      expect(actual).toMatchObject(expected)
    })

    it('should render nothing when processing work', () => {
      const ResultRendererWithJob = withJob({
        work: () => resolveAfter(workTime),
      })(ResultRenderer)
      const actual = mount(<ResultRendererWithJob />)
      expect(actual).toMatchSnapshot()
    })

    it('should render LoadingComponent when processing work', () => {
      const ResultRendererWithJob = withJob({
        work: () => resolveAfter(workTime),
        LoadingComponent,
      })(ResultRenderer)
      const actual = mount(<ResultRendererWithJob />)
      expect(actual).toMatchSnapshot()
    })

    it('should render wrapped component when work completes successfully', () => {
      const ResultRendererWithJob = withJob({
        work: () => resolveAfter(workTime, 'result'),
      })(ResultRenderer)
      const renderWrapper = mount(<ResultRendererWithJob />)
      // Allow enough time for work to complete
      return resolveAfter(workTime + 50).then(() => {
        renderWrapper.update()
        expect(renderWrapper).toMatchSnapshot()
      })
    })

    it('should render ErrorComponent when async work fails', () => {
      const error = new Error('poop')
      const ResultRendererWithJob = withJob({
        work: () => rejectAfter(workTime, error),
        ErrorComponent,
      })(ResultRenderer)
      const renderWrapper = mount(<ResultRendererWithJob />)
      // Allow enough time for work to complete
      return resolveAfter(workTime + 5 + 16).then(() => {
        renderWrapper.update()
        expect(renderWrapper).toMatchSnapshot()
      })
    })

    it('should set "error" when synchronous work fails', () => {
      const error = new Error('poop')
      const ResultRendererWithJob = withJob({
        work: () => {
          throw error
        },
        ErrorComponent,
      })(ResultRenderer)
      const renderWrapper = mount(<ResultRendererWithJob />)
      // Allow enough time for error state to be set
      return resolveAfter(16 + 1).then(() => {
        renderWrapper.update()
        expect(renderWrapper).toMatchSnapshot()
      })
    })

    it('should not fire again when no "config.shouldWorkAgain" is provided', () => {
      let fireCount = 0
      const Component = withJob({
        work: () => {
          fireCount += 1
          return 'foo'
        },
      })(() => <div>bob</div>)
      const renderWrapper = mount(<Component foo="foo" />)
      expect(fireCount).toEqual(1)
      // Set props to cause a re-render
      renderWrapper.setProps({ foo: 'bar' })
      expect(fireCount).toEqual(1)
    })

    it('should fire again for a remount', () => {
      let fireCount = 0
      const Component = withJob({
        work: () => {
          fireCount += 1
          return true
        },
      })(() => <div>bob</div>)
      mount(<Component />)
      expect(fireCount).toEqual(1)
      mount(<Component />)
      expect(fireCount).toEqual(2)
    })

    it('should fire expectantly for a "shouldWorkAgain" implementation', () => {
      const prevProductIds = []
      let fireCount = 0
      const Component = withJob({
        work: ({ productId }) => {
          prevProductIds.push(productId)
          fireCount += 1
          return true
        },
        shouldWorkAgain: (prevProps, nextProps, currentJob) =>
          (currentJob.inProgress || currentJob.completed) &&
          prevProductIds.indexOf(nextProps.productId) === -1,
      })(() => <div>bob</div>)
      const renderWrapper = mount(<Component productId={1} />)
      expect(fireCount).toEqual(1)
      renderWrapper.setProps({ productId: 2 })
      expect(fireCount).toEqual(2)
      renderWrapper.setProps({ productId: 3 })
      expect(fireCount).toEqual(3)
      renderWrapper.setProps({ productId: 2 })
      expect(fireCount).toEqual(3)
      renderWrapper.setProps({ productId: 1 })
      expect(fireCount).toEqual(3)
    })
  })
})
