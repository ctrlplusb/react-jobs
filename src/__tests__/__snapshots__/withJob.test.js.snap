// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withJob() rendering should render ErrorComponent when async work fails 1`] = `
<WithJob(ResultRenderer)>
  <ErrorComponent
    error={[Error: poop]}
  >
    <div>
      poop
    </div>
  </ErrorComponent>
</WithJob(ResultRenderer)>
`;

exports[`withJob() rendering should render LoadingComponent when processing work 1`] = `
<WithJob(ResultRenderer)>
  <LoadingComponent>
    <div>
      Loading...
    </div>
  </LoadingComponent>
</WithJob(ResultRenderer)>
`;

exports[`withJob() rendering should render nothing when processing work 1`] = `<WithJob(ResultRenderer) />`;

exports[`withJob() rendering should render wrapped component when work completes successfully 1`] = `
<WithJob(ResultRenderer)>
  <ResultRenderer
    jobResult="result"
  >
    <div>
      result
    </div>
  </ResultRenderer>
</WithJob(ResultRenderer)>
`;

exports[`withJob() rendering should set "error" when synchronous work fails 1`] = `
<WithJob(ResultRenderer)>
  <ErrorComponent
    error={[Error: poop]}
  >
    <div>
      poop
    </div>
  </ErrorComponent>
</WithJob(ResultRenderer)>
`;

exports[`withJob() rendering should set the "result" immediately if the work does not return a promise 1`] = `
<WithJob(ResultRenderer)>
  <ResultRenderer
    jobResult="bob"
  >
    <div>
      bob
    </div>
  </ResultRenderer>
</WithJob(ResultRenderer)>
`;

exports[`withJob() rendering should set the "result" immediately if the work return null 1`] = `
<WithJob(ResultRenderer)>
  <ResultRenderer
    jobResult={null}
  >
    <div />
  </ResultRenderer>
</WithJob(ResultRenderer)>
`;
